pkg/crypto/rings/cache.go

```diff
@@ -0,0 +1,201 @@
+	// used to get the addresses of the gateways an application is delegated to.
+	applicationQuerier apptypes.QueryClient
+
+	// accountQuerier is the querier for the account module, and is used to get
+	// the public keys of the application and its delegated gateways.
+	accountQuerier accounttypes.QueryClient
+}
+
+// NewRingCache returns a new RingCache instance. It requires a depinject.Config
+// to be passed in, which is used to inject the dependencies of the RingCache.
+func NewRingCache(deps depinject.Config) (RingCache, error) {
+	rc := &ringCache{
+		ringPointsCache: make(map[string][]ringtypes.Point),
+		ringPointsMu:    &sync.RWMutex{},
+	}
+
+	if err := depinject.Inject(
+		deps,
+		&rc.clientCtx,
+	); err != nil {
+		return nil, err
+	}
+
+	clientCtx := cosmosclient.Context(rc.clientCtx)
+
+	rc.accountQuerier = accounttypes.NewQueryClient(clientCtx)
+	rc.applicationQuerier = apptypes.NewQueryClient(clientCtx)
+
+	return rc, nil
+}
+
+// GetRingForAddress returns the ring for the address provided. If it does not
+// exist in the cache, it will be created by querying the application module.
+// and converting the addresses into their corresponding public key points on
+// the secp256k1 curve. It will then be cached for future use.
+func (rc *ringCache) GetRingForAddress(
+	ctx context.Context,
+	appAddress string,
+) (*ring.Ring, error) {
+	var ring *ring.Ring
+	var err error
+
+	// lock the cache for reading
+	rc.ringPointsMu.RLock()
+	// check if the ring is in the cache
+	points, ok := rc.ringPointsCache[appAddress]
+	// unlock the cache incase it was not cached
+	rc.ringPointsMu.RUnlock()
+
+	if !ok {
+		// if the ring is not in the cache, get it from the application module
+		log.Printf("DEBUG: Ring not in cache, fetching from application module [%s]", appAddress)
+		ring, err = rc.getRingForAppAddress(ctx, appAddress)
+	} else {
+		// if the ring is in the cache, create it from the points
+		log.Printf("DEBUG: Ring in cache, creating from points [%s]", appAddress)
+		ring, err = newRingFromPoints(points)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// return the ring
+	return ring, nil
+}
+
+// getRingForAppAddress returns the RingSinger used to sign relays. It does so by fetching
+// the latest information from the application module and creating the correct ring.
+// This method also caches the ring's public keys for future use.
+func (rc *ringCache) getRingForAppAddress(
+	ctx context.Context,
+	appAddress string,
+) (*ring.Ring, error) {
+	points, err := rc.getDelegatedPubKeysForAddress(ctx, appAddress)
+	if err != nil {
+		return nil, err
+	}
+	return newRingFromPoints(points)
+}
+
+// newRingFromPoints creates a new ring from points on the secp256k1 curve
+func newRingFromPoints(points []ringtypes.Point) (*ring.Ring, error) {
+	return ring.NewFixedKeyRingFromPublicKeys(ring_secp256k1.NewCurve(), points)
+}
+
+// getDelegatedPubKeysForAddress returns the ring used to sign a message for
```

Comment:
I feel like this function is doing a lot more than `getDelegatedPubKeysForAddress` and doesn't actually return pub keys but a ring.

We should either:
1. Break it out into 2 smaller helpers
2. Rename it

My preference is for (1) but not a stong opinion.

Reply:
It does return pubkeys - just in the format we desire `Points` these are then used to build the ring - they are not the ring itself